File: ExtractAnnotationFromDescription/AnnotationGroup.vb
Imports System.Collections.Generic

Friend Class AnnotationGroup
    Private m_AnnotationData As Dictionary(Of String, SortedSet(Of String))       'key=PrimaryReferenceName, value=SortedSet of xrefs
    Private m_Delimiter As String

    Sub New(GroupName As String)
        Me.GroupName = GroupName
    End Sub

    Public Property GroupName As String

    Public Property AnnotationAuthorityID As Integer

    Property XRefDelimiter As String
        Get
            Return Me.m_Delimiter
        End Get
        Set
            Me.m_Delimiter = value
        End Set
    End Property

    Public Property ImportThisGroup As Boolean

    Sub AddAnnotation(PrimaryReferenceName As String, XRefName As String)

        Dim xrefList As SortedSet(Of String)

        If Me.m_AnnotationData Is Nothing Then
            Me.m_AnnotationData = New Dictionary(Of String, SortedSet(Of String))
        End If

        If Not Me.m_AnnotationData.ContainsKey(PrimaryReferenceName) Then
            xrefList = New SortedSet(Of String)
            xrefList.Add(XRefName)
        Else
            xrefList = m_AnnotationData.Item(PrimaryReferenceName.ToString)
            If Not xrefList.Contains(XRefName) Then
                xrefList.Add(XRefName)
                Me.m_AnnotationData.Item(PrimaryReferenceName.ToString) = xrefList
            End If
        End If
    End Sub

    Function GetAllXRefs() As Dictionary(Of String, SortedSet(Of String))
        Return Me.m_AnnotationData
    End Function

    Function GetAllPrimaryReferences() As SortedSet(Of String)
        Dim s As String
        Dim annotationKeys As New SortedSet(Of String)
        For Each s In Me.m_AnnotationData.Keys
            annotationKeys.Add(s)
        Next

        Return annotationKeys
    End Function

    Function GetXRefs(PrimaryReferenceName As String) As SortedSet(Of String)
        Dim xrefList = Me.m_AnnotationData.Item(PrimaryReferenceName)

        If Me.m_Delimiter.Length > 0 Then
            Dim addnXRefs() As String
            Dim primeXRef As String
            Dim XRefCount As Integer
            Dim newXReflist As New SortedSet(Of String)

            For Each primeXRef In xrefList
                addnXRefs = primeXRef.Split(Me.m_Delimiter.ToCharArray)
                For XRefCount = 0 To addnXRefs.Length - 1
                    Dim newItem = addnXRefs(XRefCount).ToString
                    If Not newXReflist.Contains(newItem) Then
                        newXReflist.Add(newItem)
                    End If
                Next
            Next

            xrefList = newXReflist

        End If

        Return xrefList
    End Function
End Class

File: ExtractAnnotationFromDescription/AnnotationGroup.cs
using System.Collections.Generic;

namespace ExtractAnnotationFromDescription
{
    internal class AnnotationGroup
    {
        private Dictionary<string, SortedSet<string>> m_AnnotationData;       // key=PrimaryReferenceName, value=SortedSet of xrefs
        private string m_Delimiter;

        public AnnotationGroup(string GroupName)
        {
            this.GroupName = GroupName;
        }

        public string GroupName { get; set; }

        public int AnnotationAuthorityID { get; set; }

        public string XRefDelimiter
        {
            get => m_Delimiter;
            set => m_Delimiter = value;
        }

        public bool ImportThisGroup { get; set; }

        public void AddAnnotation(string PrimaryReferenceName, string XRefName)
        {
            SortedSet<string> xrefList;

            if (m_AnnotationData == null)
            {
                m_AnnotationData = new Dictionary<string, SortedSet<string>>();
            }

            if (!m_AnnotationData.ContainsKey(PrimaryReferenceName))
            {
                xrefList = new SortedSet<string>();
                xrefList.Add(XRefName);
            }
            else
            {
                xrefList = m_AnnotationData[PrimaryReferenceName.ToString()];
                if (!xrefList.Contains(XRefName))
                {
                    xrefList.Add(XRefName);
                    m_AnnotationData[PrimaryReferenceName.ToString()] = xrefList;
                }
            }
        }

        public Dictionary<string, SortedSet<string>> GetAllXRefs()
        {
            return m_AnnotationData;
        }

        public SortedSet<string> GetAllPrimaryReferences()
        {
            var annotationKeys = new SortedSet<string>();
            foreach (var s in m_AnnotationData.Keys)
                annotationKeys.Add(s);

            return annotationKeys;
        }

        public SortedSet<string> GetXRefs(string PrimaryReferenceName)
        {
            var xrefList = m_AnnotationData[PrimaryReferenceName];

            if (m_Delimiter.Length > 0)
            {
                string[] addnXRefs;
                int XRefCount;
                var newXReflist = new SortedSet<string>();

                foreach (var primeXRef in xrefList)
                {
                    addnXRefs = primeXRef.Split(m_Delimiter.ToCharArray());
                    for (XRefCount = 0; XRefCount < addnXRefs.Length; XRefCount++)
                    {
                        string newItem = addnXRefs[XRefCount].ToString();
                        if (!newXReflist.Contains(newItem))
                        {
                            newXReflist.Add(newItem);
                        }
                    }
                }

                xrefList = newXReflist;
            }

            return xrefList;
        }
    }
}